From 37f20e45e126929c8273f3d6b0c5b20d9e8ba7f9 Mon Sep 17 00:00:00 2001
From: Aaron Kling <webgeek1234@gmail.com>
Date: Thu, 17 Nov 2016 15:53:03 -0600
Subject: [PATCH] zygote: Allow device to append extra whitelisted paths

Since this was approved for cm-13.0, it needs put back in here.

Change-Id: Ibbffdced9ee86370821e4c8766a8dc9fe6a6af99
---
 core/jni/fd_utils-inl-extra.h | 30 +++++++++++++++++++++++++++++
 core/jni/fd_utils.cpp         |  5 +++++
 2 files changed, 35 insertions(+)
 create mode 100644 core/jni/fd_utils-inl-extra.h

diff --git a/core/jni/fd_utils-inl-extra.h b/core/jni/fd_utils-inl-extra.h
new file mode 100644
index 00000000000..993c320897e
--- /dev/null
+++ b/core/jni/fd_utils-inl-extra.h
@@ -0,0 +1,29 @@
+/*
+ * Copyright (C) 2016 The CyanogenMod Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+#define PATH_WHITELIST_EXTRA_H \
+    "/proc/ged", \
+    "/proc/apid", \
+    "/proc/aprf",
+*/
+
+// Overload this file in your device specific config if you need
+// to add extra whitelisted paths.
+// WARNING: Only use this if necessary. Custom inits should be
+// checked for leaked file descriptors before even considering
+// this.
+// In order to add your files, copy the whole file (don't forget the copyright notice!),
+// uncomment the #define above and change the paths inside to match your requirements
\ No newline at end of file
diff --git a/core/jni/fd_utils.cpp b/core/jni/fd_utils.cpp
index 92a542676bc..d8519f4d734 100644
--- a/core/jni/fd_utils.cpp
+++ b/core/jni/fd_utils.cpp
@@ -31,6 +31,8 @@
 #include <android-base/stringprintf.h>
 #include <android-base/strings.h>
 
+#include <fd_utils-inl-extra.h>
+
 // Static whitelist of open paths that the zygote is allowed to keep open.
 static const char* kPathWhitelist[] = {
   "/dev/null",
@@ -43,6 +45,9 @@ static const char* kPathWhitelist[] = {
   "/dev/ion",
   "/dev/dri/renderD129", // Fixes b/31172436
   "/system/framework/org.lineageos.platform-res.apk",
+#ifdef PATH_WHITELIST_EXTRA_H
+PATH_WHITELIST_EXTRA_H
+#endif
 };
 
 static const char kFdPath[] = "/proc/self/fd";
-- 
2.11.0

